File: scripts/plot_genotype_blocks.py
#!/usr/bin/env python3
"""Per-block genotyping error, one figure per locus.

Two rows: leave-zero-out above (the control -- the donor's own haplotypes are IN the panel, so a
perfect model reads 0), leave-one-out below (the donor is held out, so part of the error is a
haplotype the panel no longer contains). One violin pair per block, real reads against simulated.

Metric is normalized edit distance to truth, 1 - identity, so blocks of very different length share
an axis. The same metric is used in both rows on purpose: mixing "distance to truth" above with
"distance to the best available pair" below would make the two rows incomparable.

The panel floor (1 - best_identity) is drawn as a black diamond. Under leave-one-out the violin
ABOVE that diamond is the model's own error; the diamond is what removing the donor costs. Plotting
the violin alone would charge the model for a haplotype nobody could have picked.

NOTE `best_identity` is computed over a top-16 syncmer-Jaccard shortlist, so it is a LOWER bound on
the true panel ceiling: the real floor may be lower and the model's own share correspondingly larger.

  plot_genotype_blocks.py --table all_blocks.tsv --out DIR [--loci c4,lpa] [--dpi 150]
"""
import argparse, os, sys
import numpy as np
import pandas as pd
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
from matplotlib.patches import Patch

PAL = {'real': '#B2182B', 'simulated': '#2166AC'}
ROWS = [('LZO', 'leave-zero-out  (donor kept in panel)'),
        ('LOO', 'leave-one-out  (donor removed)')]

ap = argparse.ArgumentParser(description=__doc__,
                             formatter_class=argparse.RawDescriptionHelpFormatter)
ap.add_argument('--table', required=True, help='per-block cohort table')
ap.add_argument('--out', required=True, help='output directory')
ap.add_argument('--loci', default=None, help='comma-separated subset')
ap.add_argument('--dpi', type=int, default=150)
a = ap.parse_args()

d = pd.read_csv(a.table, sep='\t')
need = {'locus', 'read_source', 'regime', 'block_index', 'block_kind',
        'sample', 'identity', 'best_identity'}
miss = need - set(d.columns)
if miss:
    sys.exit(f'table lacks columns: {", ".join(sorted(miss))}')

d['identity'] = pd.to_numeric(d['identity'], errors='coerce')
d['best_identity'] = pd.to_numeric(d['best_identity'], errors='coerce')
d = d[np.isfinite(d['identity'])].copy()
d['err'] = 1.0 - d['identity']
d['floor'] = 1.0 - d['best_identity']
if 'exact' in d.columns:
    d['exact'] = pd.to_numeric(d['exact'], errors='coerce')
else:
    d['exact'] = np.nan
if a.loci:
    d = d[d['locus'].isin(a.loci.split(','))]
if d.empty:
    sys.exit('no rows to plot')

os.makedirs(a.out, exist_ok=True)

# Per-block summary written alongside, since "what percent did we genotype properly" is a number
# people want to quote and reading it off a violin is guesswork.
rows = []
for (loc, reg, blk, kind, src), g in d.groupby(
        ['locus', 'regime', 'block_index', 'block_kind', 'read_source'], sort=True):
    rows.append(dict(locus=loc, regime=reg, block_index=blk, block_kind=kind, read_source=src,
                     n=len(g), exact_pct=100.0 * g['exact'].mean(skipna=True),
                     mean_err=g['err'].mean(), median_err=g['err'].median(),
                     max_err=g['err'].max(), panel_floor=g['floor'].mean(skipna=True)))
summary = pd.DataFrame(rows).sort_values(['locus', 'regime', 'block_index', 'read_source'])
summary.to_csv(os.path.join(a.out, 'summary.tsv'), sep='\t', index=False,
               float_format='%.6f')

for loc in sorted(d['locus'].unique()):
    sub = d[d['locus'] == loc]
    blocks = sorted(sub['block_index'].unique())
    kind = {b: str(sub.loc[sub['block_index'] == b, 'block_kind'].iloc[0]) for b in blocks}
    fig, axes = plt.subplots(2, 1, figsize=(max(7.0, 0.46 * len(blocks) + 2.6), 7.2), sharex=True)

    for ax, (reg, title) in zip(axes, ROWS):
        r = sub[sub['regime'] == reg]
        for off, src in ((-0.19, 'real'), (0.19, 'simulated')):
            xs, data = [], []
            for i, b in enumerate(blocks):
                v = r.loc[(r['block_index'] == b) & (r['read_source'] == src), 'err'].dropna()
                if len(v) == 0:
                    continue
                # A block where every sample is exact has zero variance; violin() fails on that, so
                # those are drawn as a flat marker rather than silently dropped.
                if len(v) > 1 and v.nunique() > 1:
                    xs.append(i + off); data.append(v.values)
                else:
                    ax.plot([i + off - 0.13, i + off + 0.13], [v.iloc[0]] * 2,
                            color=PAL[src], lw=1.6, solid_capstyle='butt', zorder=3)
                jit = (np.random.default_rng(42).random(len(v)) - 0.5) * 0.16
                ax.scatter(i + off + jit, v.values, s=3, color=PAL[src], alpha=0.45,
                           linewidths=0, zorder=4)
            if data:
                parts = ax.violinplot(data, positions=xs, widths=0.34, showextrema=False)
                for pc in parts['bodies']:
                    pc.set_facecolor(PAL[src]); pc.set_alpha(0.45); pc.set_edgecolor('none')
        fl = [r.loc[r['block_index'] == b, 'floor'].mean() for b in blocks]
        ax.scatter(range(len(blocks)), fl, marker='D', s=14, color='black', zorder=5,
                   label='panel floor (1 - best_identity)')
        ax.set_title(title, fontsize=10, loc='left', fontweight='bold')
        ax.set_ylabel('1 - identity')
        ax.grid(axis='y', lw=0.4, alpha=0.35)
        ax.set_axisbelow(True)
        ax.margins(x=0.01)

    axes[1].set_xticks(range(len(blocks)))
    axes[1].set_xticklabels([f'{b}\n{kind[b][:4]}' for b in blocks], fontsize=7)
    axes[1].set_xlabel('block (chain order; kind abbreviated)')
    handles = [Patch(facecolor=PAL['real'], alpha=0.6, label='real reads'),
               Patch(facecolor=PAL['simulated'], alpha=0.6, label='simulated reads'),
               plt.Line2D([], [], marker='D', color='black', ls='none', ms=4,
                          label='panel floor (unreachable under LOO)')]
    axes[0].legend(handles=handles, loc='upper right', fontsize=7, framealpha=0.9)
    ns = sub.groupby('read_source')['sample'].nunique().to_dict()
    fig.suptitle(f'{loc} - per-block genotyping error   '
                 f'({ns.get("real", 0)} real / {ns.get("simulated", 0)} simulated donors)',
                 fontsize=12, fontweight='bold')
    fig.tight_layout(rect=(0, 0, 1, 0.97))
    p = os.path.join(a.out, f'{loc}.png')
    fig.savefig(p, dpi=a.dpi); plt.close(fig)
    print(f'wrote {p}  ({len(blocks)} blocks, {len(sub)} observations)')

print(f'wrote {os.path.join(a.out, "summary.tsv")}')

File: scripts/plot_genotype_blocks.R
#!/usr/bin/env Rscript
# Per-block genotyping error, one plot per locus: two rows (leave-zero-out above, leave-one-out
# below), one violin pair per block (real reads against simulated), blocks in chain order.
#
# The metric is NORMALIZED EDIT DISTANCE to truth, 1 - identity, so blocks of wildly different length
# sit on one axis. The same metric is used in both rows, which is the point: LZO is the control where
# the answer is in the panel and the error should be ~0, and any distance from 0 there is read
# acquisition or model error rather than an unrepresentable truth.
#
# LOO carries an irreducible part -- the individual's own haplotypes are removed, so even a perfect
# model can only reach the nearest remaining panel pair. That floor (1 - best_identity) is drawn as a
# black point per block, so the violin ABOVE the point is the model's own error and the point itself
# is what the panel costs. Reporting the violin alone would blame the model for a missing haplotype.
#
#   Rscript plot_genotype_blocks.R --table all_blocks.tsv --out DIR [--loci c4,lpa] [--dpi 150]
suppressWarnings(suppressMessages({library(ggplot2)}))

args <- commandArgs(trailingOnly = TRUE)
usage <- function(status = 0) {
  cat(paste(c(
    "plot_genotype_blocks.R - per-block genotyping error, one plot per locus.",
    "",
    "  Two rows per locus: leave-zero-out (truth in the panel) above, leave-one-out below.",
    "  One violin pair per block: real reads against simulated, on normalized edit distance",
    "  to truth (1 - identity). The black point marks the panel floor (1 - best_identity),",
    "  which is what remains unreachable once the individual's haplotypes are removed.",
    "",
    "Usage:",
    "  Rscript plot_genotype_blocks.R --table <all_blocks.tsv> --out <dir> [options]",
    "",
    "Required:",
    "  --table <f>     per-block cohort table with locus, read_source, regime, block_index,",
    "                  block_kind, sample, identity, best_identity, exact",
    "  --out <dir>     output directory; writes <dir>/<locus>.png and <dir>/summary.tsv",
    "",
    "Options:",
    "  --loci <a,b>    only these loci (default: all present)",
    "  --dpi <n>       raster resolution (default 150)",
    "  --width <in>    plot width; default scales with the block count",
    "  -h, --help      this message",
    ""), collapse = "\n"))
  quit(status = status)
}
if (length(args) == 0 || "-h" %in% args || "--help" %in% args) usage()

getarg <- function(flag, default = NA) {
  i <- match(flag, args)
  if (is.na(i)) return(default)
  if (i == length(args)) stop(paste(flag, "needs a value"))
  args[i + 1]
}
table_path <- getarg("--table")
out_dir    <- getarg("--out")
if (is.na(table_path) || is.na(out_dir)) usage(1)
if (!file.exists(table_path)) stop(paste("no such table:", table_path))
dpi   <- as.numeric(getarg("--dpi", 150))
width_override <- suppressWarnings(as.numeric(getarg("--width", NA)))
only_loci <- getarg("--loci", NA)

d <- read.delim(table_path, stringsAsFactors = FALSE, check.names = FALSE)
need <- c("locus", "read_source", "regime", "block_index", "block_kind",
          "sample", "identity", "best_identity", "filter")
missing <- setdiff(need, names(d))
if (length(missing)) stop(paste("table lacks columns:", paste(missing, collapse = ", ")))

d$identity      <- suppressWarnings(as.numeric(d$identity))
d$best_identity <- suppressWarnings(as.numeric(d$best_identity))
d <- d[is.finite(d$identity), ]
if (!nrow(d)) stop("no rows with a numeric identity")

# Distance to truth, and the part of it the panel cannot avoid once the donor is held out.
d$err   <- 1 - d$identity
d$floor <- ifelse(is.finite(d$best_identity), 1 - d$best_identity, NA_real_)

if (!is.na(only_loci)) {
  keep <- strsplit(only_loci, ",")[[1]]
  d <- d[d$locus %in% keep, ]
  if (!nrow(d)) stop("no rows for the requested loci")
}

d$regime <- factor(ifelse(d$regime %in% c("LZO", "0"), "leave-zero-out (truth in panel)",
                          "leave-one-out (donor removed)"),
                   levels = c("leave-zero-out (truth in panel)", "leave-one-out (donor removed)"))
d$reads <- factor(d$read_source, levels = c("real", "simulated"))
d <- d[!is.na(d$reads), ]

dir.create(out_dir, showWarnings = FALSE, recursive = TRUE)
pal <- c(real = "#B2182B", simulated = "#2166AC")

# Per (locus, regime, block, reads): the exact rate is what "genotyped properly" means -- the called
# allele pair IS the truth pair. Kept next to the error so a block that is usually exact but
# occasionally catastrophic is not confused with one that is uniformly mediocre.
if ("exact" %in% names(d)) {
  d$exact <- suppressWarnings(as.numeric(d$exact))
} else {
  d$exact <- NA_real_
}
agg <- aggregate(cbind(err, exact, floor) ~ locus + regime + block_index + block_kind + reads,
                 data = d, FUN = function(x) mean(x, na.rm = TRUE), na.action = na.pass)
names(agg)[names(agg) == "err"] <- "mean_err"
n_obs <- aggregate(err ~ locus + regime + block_index + reads, data = d, FUN = length)
names(n_obs)[names(n_obs) == "err"] <- "n"
agg <- merge(agg, n_obs, by = c("locus", "regime", "block_index", "reads"))
agg <- agg[order(agg$locus, agg$regime, agg$block_index, agg$reads), ]
write.table(agg, file.path(out_dir, "summary.tsv"), sep = "\t", quote = FALSE, row.names = FALSE)

for (loc in sort(unique(d$locus))) {
  sub <- d[d$locus == loc, ]
  blocks <- sort(unique(sub$block_index))
  # Label each block with its kind, since "which block is the bubble" is the first thing asked.
  kind_of <- sapply(blocks, function(b) {
    k <- unique(sub$block_kind[sub$block_index == b]); k[1]
  })
  lab <- paste0(blocks, "\n", substr(kind_of, 1, 4))
  sub$blk <- factor(sub$block_index, levels = blocks, labels = lab)

  fl <- unique(sub[, c("regime", "blk", "floor", "block_index")])
  fl <- fl[is.finite(fl$floor), ]
  fl <- aggregate(floor ~ regime + blk, data = fl, FUN = mean)

  # Which of this error the module already declines. A reader has to be able to tell "we call this
  # wrong" from "we decline to call this", because only the first is a defect -- the second is the
  # module correctly saying the block cannot be genotyped on the evidence it has.
  sub$kept <- ifelse(sub$filter == "PASS", "reported (PASS)", "declined")
  sub$kept <- factor(sub$kept, levels = c("reported (PASS)", "declined"))

  p <- ggplot(sub, aes(x = blk, y = err, fill = reads)) +
    # Violins collapse to a line when every sample is exact, which is the correct picture for a
    # solved block; the jittered points keep the sample count visible when that happens.
    geom_violin(position = position_dodge(width = 0.8), width = 0.75,
                scale = "width", alpha = 0.55, colour = NA, na.rm = TRUE) +
    geom_point(position = position_jitterdodge(jitter.width = 0.12, dodge.width = 0.8,
                                               seed = 42),
               aes(colour = reads, shape = kept, size = kept, alpha = kept), na.rm = TRUE) +
    scale_shape_manual(values = c("reported (PASS)" = 16, "declined" = 1), name = NULL) +
    scale_size_manual(values = c("reported (PASS)" = 0.75, "declined" = 0.9), guide = "none") +
    scale_alpha_manual(values = c("reported (PASS)" = 0.75, "declined" = 0.45), guide = "none") +
    geom_point(data = fl, aes(x = blk, y = floor), inherit.aes = FALSE,
               shape = 18, size = 1.9, colour = "black", na.rm = TRUE) +
    facet_wrap(~ regime, ncol = 1, scales = "free_y") +
    scale_fill_manual(values = pal, name = "reads") +
    scale_colour_manual(values = pal, guide = "none") +
    labs(title = paste0(loc, " - per-block genotyping error"),
         subtitle = paste0("normalized edit distance to truth (1 - identity). Filled = reported, ",
                           "hollow = declined by the module. Black diamond = panel floor ",
                           "(1 - best_identity), unreachable once the donor is held out."),
         x = "block (chain order, kind abbreviated)", y = "1 - identity") +
    theme_bw(base_size = 9) +
    theme(panel.grid.minor = element_blank(),
          strip.background = element_rect(fill = "grey92", colour = NA),
          strip.text = element_text(face = "bold"),
          legend.position = "top",
          plot.subtitle = element_text(size = 7, colour = "grey30"))

  w <- if (is.finite(width_override)) width_override else max(6, 0.42 * length(blocks) + 2.2)
  ggsave(file.path(out_dir, paste0(loc, ".png")), p,
         width = w, height = 6.0, dpi = dpi, limitsize = FALSE)
  cat(sprintf("wrote %s (%d blocks, %d observations)\n",
              file.path(out_dir, paste0(loc, ".png")), length(blocks), nrow(sub)))
}
